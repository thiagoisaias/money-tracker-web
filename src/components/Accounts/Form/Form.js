import React, { Component } from "react";
import PropTypes from "prop-types";

import {
  Container,
  Row,
  Message,
  ColoredMark,
  Title,
  FormContainer,
  FormGroup,
  Label,
  SubmitButton,
  ErrorMessage
} from "components/UI/Form/Form";

import Input from "components/UI/Input/Input";
import Layout from "components/Layout/Layout";
import Spinner from "components/UI/Spinner/Spinner";
import withFormHandler from "hoc/withFormHandler/withFormHandler";

class Form extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isFormValid: false,
      formFields: {
        name: {
          elementType: "input",
          elementConfig: {
            type: "text",
            placeholder: "Enter account name"
          },
          name: "Name",
          value: this.props.name || "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        initialBalance: {
          elementType: "input",
          elementConfig: {
            type: "number",
            placeholder: "Enter initial balance"
          },
          name: "Initial Balance",
          value: this.props.initialBalance || "",
          validation: {
            required: true,
            nonNegative: true
          },
          isValid: false,
          touched: false
        }
      }
    };
  }

  checkFieldValidity = (value, rules) => {
    let isValid = true;

    if (rules.required) {
      isValid = value.trim() !== "" && isValid;
    }

    if (rules.nonNegative) {
      isValid = value >= 0 && isValid;
    }

    return isValid;
  };

  onInputChange = event => {
    const { handleInputChange } = this.props;
    const formFields = { ...this.state.formFields };
    const checkFieldValidity = this.checkFieldValidity;
    event.preventDefault();
    const updatedFormState = handleInputChange(
      event,
      formFields,
      checkFieldValidity
    );
    this.setState({
      ...updatedFormState
    });
  };

  handleSubmit = event => {
    const { generateFormData, onSubmitData } = this.props;
    event.preventDefault();

    const formData = generateFormData(this.state.formFields);

    onSubmitData(formData);
  };

  componentDidMount() {
    const { match, checkFormValidity } = this.props;

    const isFormValid = checkFormValidity(
      this.state.formFields,
      this.checkFieldValidity
    );

    if (match.path === "/accounts/:id/edit" && isFormValid) {
      this.setState({
        ...this.state,
        isFormValid: true
      });
    }
  }

  render() {
    const { error, isLoading, match } = this.props;
    const formTitle =
      match.path === "/accounts/new" ? "Add Account" : "Edit Account";

    const keyList = [];
    for (let key in this.state.formFields) {
      keyList.push(key);
    }

    const inputList = keyList.map(key => {
      const formField = this.state.formFields[key];
      return (
        <FormGroup key={key}>
          <Label>
            {formField.name} <ColoredMark>*</ColoredMark>
          </Label>
          <Input
            elementConfig={formField.elementConfig}
            elementType={formField.elementType}
            name={key}
            value={formField.value}
            touched={formField.touched}
            isValid={formField.isValid}
            onChange={this.onInputChange}
          />
        </FormGroup>
      );
    });

    return (
      <Layout>
        <Container>
          <Title>{formTitle}</Title>
          {error && <ErrorMessage>{error}</ErrorMessage>}
          <FormContainer onSubmit={this.handleSubmit}>
            {inputList}
            <Row>
              <SubmitButton
                disabled={!this.state.isFormValid}
                onClick={this.handleSubmit}
              >
                {isLoading ? <Spinner size={25} color={"#ddd"} /> : "Submit"}
              </SubmitButton>
              <Message>
                <ColoredMark>*</ColoredMark>
                {" represent required fields."}
              </Message>
            </Row>
          </FormContainer>
        </Container>
      </Layout>
    );
  }
}

Form.propTypes = {
  checkFormValidity: PropTypes.func.isRequired,
  id: PropTypes.number,
  name: PropTypes.string,
  error: PropTypes.string,
  generateFormData: PropTypes.func.isRequired,
  handleInputChange: PropTypes.func.isRequired,
  initialBalance: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  match: PropTypes.object.isRequired,
  onSubmitData: PropTypes.func.isRequired
};

export default withFormHandler(Form);
