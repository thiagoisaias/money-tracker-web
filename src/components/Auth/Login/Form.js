import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";

import {
  StyledForm,
  SubmitButton,
  Title,
  Alternate,
  Link,
  ErrorMessage
} from "./styled";

import Input from "components/UI/Input/Input";
import Spinner from "components/UI/Spinner/Spinner";
import withFormHandler from "hoc/withFormHandler/withFormHandler";

class Form extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isFormValid: false,
      formFields: {
        email: {
          elementType: "input",
          elementConfig: {
            type: "text",
            placeholder: "Email",
            autoComplete: "email"
          },
          name: "Email",
          value: this.props.email || "",
          validation: {},
          isValid: false,
          touched: false
        },
        password: {
          elementType: "input",
          elementConfig: {
            type: "password",
            placeholder: "Password",
            autoComplete: "password"
          },
          name: "Password",
          value: this.props.name || "",
          validation: {},
          isValid: false,
          touched: false
        }
      }
    };
  }

  checkFieldValidity = (value, rules) => {
    let isValid = true;

    // TODO: Add email validation

    return isValid;
  };

  onInputChange = event => {
    const { handleInputChange } = this.props;
    const formFields = { ...this.state.formFields };
    const checkFieldValidity = this.checkFieldValidity;
    const updatedFormState = handleInputChange(
      event,
      formFields,
      checkFieldValidity
    );
    this.setState({
      ...updatedFormState
    });
  };

  handleSubmit = event => {
    const { onSubmitData } = this.props;
    event.preventDefault();

    const formData = {};

    for (let fieldKey in this.state.formFields) {
      formData[fieldKey] = this.state.formFields[fieldKey].value;
    }

    onSubmitData(formData);
  };

  render() {
    const { isLoading, error } = this.props;

    const keyList = [];
    for (let key in this.state.formFields) {
      keyList.push(key);
    }

    const inputList = keyList.map(key => {
      const formField = this.state.formFields[key];
      return (
        <Input
          key={key}
          {...formField.elementConfig}
          name={key}
          value={formField.value}
          touched={formField.touched}
          isValid={formField.isValid}
          onChange={this.onInputChange}
        />
      );
    });

    return (
      <Fragment>
        <Title>{"Sign in"}</Title>
        {error && <ErrorMessage> {error} </ErrorMessage>}
        <StyledForm onSubmit={this.handleSubmit}>
          {inputList}
          <SubmitButton onClick={this.handleSubmit}>
            {isLoading ? <Spinner size={25} color={"#ddd"} /> : "Log In"}
          </SubmitButton>
        </StyledForm>
        <Alternate>
          {"DoesnÂ´t have an account? "}
          <Link to="/auth/signup">{"Sign up here"}</Link>
        </Alternate>
      </Fragment>
    );
  }
}

Form.propTypes = {
  error: PropTypes.string,
  handleInputChange: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  onSubmitData: PropTypes.func.isRequired
};

export default withFormHandler(Form);
