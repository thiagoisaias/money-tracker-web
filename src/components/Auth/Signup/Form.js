import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";

import {
  StyledForm,
  SubmitButton,
  Title,
  Alternate,
  Link,
  ErrorMessage
} from "./styled";

import Input from "components/UI/Input/Input";
import Spinner from "components/UI/Spinner/Spinner";
import withFormHandler from "hoc/withFormHandler/withFormHandler";

class Form extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isFormValid: false,
      formFields: {
        name: {
          elementType: "input",
          elementConfig: {
            type: "text",
            placeholder: "Full name",
            autoComplete: "name"
          },
          name: "Name",
          value: "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        email: {
          elementType: "input",
          elementConfig: {
            type: "text",
            placeholder: "Email",
            autoComplete: "email"
          },
          name: "Email",
          value: "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        password: {
          elementType: "input",
          elementConfig: {
            type: "password",
            placeholder: "Password",
            autoComplete: "password"
          },
          name: "Password",
          value: "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        passwordConfirmation: {
          elementType: "input",
          elementConfig: {
            type: "password",
            placeholder: "Password confirmation",
            autoComplete: "passwordConfirmation"
          },
          name: "Password",
          value: "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        }
      }
    };
  }

  checkFieldValidity = (value, rules) => {
    let isValid = true;

    if (rules.required) {
      isValid = value.trim() !== "" && isValid;
    }

    return isValid;
  };

  onInputChange = event => {
    const { handleInputChange } = this.props;
    const formFields = { ...this.state.formFields };
    const checkFieldValidity = this.checkFieldValidity;
    const updatedFormState = handleInputChange(
      event,
      formFields,
      checkFieldValidity
    );
    this.setState({
      ...updatedFormState
    });
  };

  handleSubmit = event => {
    const { onSubmitData } = this.props;
    event.preventDefault();

    const formData = {};

    for (let fieldKey in this.state.formFields) {
      formData[fieldKey] = this.state.formFields[fieldKey].value;
    }

    onSubmitData(formData);
  };

  render() {
    const { isLoading, error } = this.props;

    const keyList = [];
    for (let key in this.state.formFields) {
      keyList.push(key);
    }

    const inputList = keyList.map(key => {
      const formField = this.state.formFields[key];
      return (
        <Input
          key={key}
          {...formField.elementConfig}
          name={key}
          value={formField.value}
          touched={formField.touched}
          isValid={formField.isValid}
          onChange={this.onInputChange}
        />
      );
    });

    return (
      <Fragment>
        <Title>{"Create an account"}</Title>
        <StyledForm onSubmit={this.handleSubmit}>
          {error && <ErrorMessage> {error} </ErrorMessage>}
          {inputList}
          <SubmitButton onClick={this.handleSubmit}>
            {isLoading ? <Spinner size={25} color={"#ddd"} /> : "Sign Up"}
          </SubmitButton>
        </StyledForm>
        <Alternate>
          {"Already have an account? "}
          <Link to="/auth/login">{"Login here"}</Link>
        </Alternate>
      </Fragment>
    );
  }
}

Form.propTypes = {
  error: PropTypes.string,
  handleInputChange: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  onSubmitData: PropTypes.func.isRequired
};

export default withFormHandler(Form);
