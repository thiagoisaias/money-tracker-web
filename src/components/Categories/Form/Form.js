import React, { Component } from "react";
import PropTypes from "prop-types";

import {
  Container,
  StyledInput,
  Row,
  Message,
  ColoredMark,
  Title,
  FormContainer,
  FormGroup,
  Label,
  SubmitButton,
  ErrorMessage
} from "shared/Form/styled";

import Layout from "components/Layout/Layout";
import Spinner from "shared/Spinner/Spinner";
import withFormHandler from "hoc/withFormHandler/withFormHandler";

class Form extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isFormValid: false,
      formFields: {
        name: {
          elementType: "input",
          elementConfig: {
            type: "text",
            placeholder: "Enter category name"
          },
          name: "Name",
          value: this.props.name || "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        color: {
          elementType: "input",
          elementConfig: {
            type: "text",
            placeholder: "Enter color"
          },
          name: "Color",
          value: this.props.color || "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        }
      }
    };
  }

  checkFieldValidity = (value, rules) => {
    let isValid = true;

    if (rules.required) {
      isValid = value.trim() !== "" && isValid;
    }

    return isValid;
  };

  onInputChange = event => {
    const { handleInputChange } = this.props;
    const formFields = { ...this.state.formFields };
    const checkFieldValidity = this.checkFieldValidity;
    event.preventDefault();
    const updatedFormState = handleInputChange(
      event,
      formFields,
      checkFieldValidity
    );
    this.setState({
      ...updatedFormState
    });
  };

  handleSubmit = event => {
    const { generateFormData, onSubmitData } = this.props;
    event.preventDefault();

    const formData = generateFormData(this.state.formFields);

    onSubmitData(formData);
  };

  componentDidMount() {
    const { match, checkFormValidity } = this.props;

    const isFormValid = checkFormValidity(
      this.state.formFields,
      this.checkFieldValidity
    );

    if (match.path === "/categories/:id/edit" && isFormValid) {
      this.setState({
        ...this.state,
        isFormValid: true
      });
    }
  }

  render() {
    const { error, isLoading, match } = this.props;
    const formTitle =
      match.path === "/categories/new" ? "Add Category" : "Edit Category";

    const keyList = [];
    for (let key in this.state.formFields) {
      keyList.push(key);
    }

    const inputList = keyList.map(key => {
      const formField = this.state.formFields[key];
      return (
        <FormGroup key={key}>
          <Label>
            {formField.name} <ColoredMark>*</ColoredMark>
          </Label>
          <StyledInput
            {...formField.elementConfig}
            value={formField.value}
            touched={formField.touched}
            isValid={formField.isValid}
            onChange={this.onInputChange}
          />
        </FormGroup>
      );
    });

    return (
      <Layout>
        <Container>
          <Title>{formTitle}</Title>
          {error && <ErrorMessage>{error}</ErrorMessage>}
          <FormContainer onSubmit={this.handleSubmit}>
            {inputList}
            <Row>
              <SubmitButton
                disabled={!this.state.isFormValid}
                onClick={this.handleSubmit}
              >
                {isLoading ? <Spinner size={25} color={"#ddd"} /> : "Submit"}
              </SubmitButton>
              <Message>
                <ColoredMark>*</ColoredMark>
                {" represent required fields."}
              </Message>
            </Row>
          </FormContainer>
        </Container>
      </Layout>
    );
  }
}

Form.propTypes = {
  checkFormValidity: PropTypes.func.isRequired,
  id: PropTypes.number,
  color: PropTypes.string,
  error: PropTypes.string,
  generateFormData: PropTypes.func.isRequired,
  handleInputChange: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  match: PropTypes.object.isRequired,
  name: PropTypes.string,
  onSubmitData: PropTypes.func.isRequired
};

export default withFormHandler(Form);
