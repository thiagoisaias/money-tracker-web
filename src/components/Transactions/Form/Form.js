import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import moment from "moment";

import {
  Wrapper,
  SelectWrapper,
  InputWrapper,
  SubmitRow,
  Row,
  StyledDatePicker,
  Message,
  ColoredMark,
  Title,
  FormContainer,
  Label,
  SubmitButton,
  ErrorMessage
} from "./styled";

import Input from "shared/Input/Input";
import Layout from "components/Layout/Layout";
import Spinner from "shared/Spinner/Spinner";
import withFormHandler from "hoc/withFormHandler/withFormHandler";

class Form extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isFormValid: false,
      formFields: {
        transactionType: {
          elementType: "select",
          elementConfig: {
            options: [
              { value: "expense", label: "Expense" },
              { value: "earning", label: "Earning" }
            ]
          },
          label: "Type",
          value: "expense",
          validation: {},
          isValid: false,
          touched: false
        },
        description: {
          elementType: "input",
          elementConfig: {
            type: "text",
            placeholder: "Add a transaction description"
          },
          label: "Description",
          value: this.props.description || "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        value: {
          elementType: "input",
          elementConfig: {
            type: "number",
            placeholder: "$ 0.00"
          },
          label: "Value",
          value: this.props.value || "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        date: {
          elementType: "input",
          elementConfig: {
            type: "date"
          },
          label: "Date",
          value: this.props.date || moment(),
          validation: {
            requiredDate: true
          },
          isValid: false,
          touched: false
        },
        account: {
          elementType: "select",
          elementConfig: {
            options: [{ id: 1, label: "ole" }, { id: 2, label: "ola" }]
          },
          label: "Account",
          value: 1,
          validation: {},
          isValid: true,
          touched: false
        },
        category: {
          elementType: "select",
          elementConfig: {
            options: [{ id: 1, label: "ole" }, { id: 2, label: "ola" }]
          },
          label: "Category",
          value: 1,
          validation: {},
          isValid: true,
          touched: false
        }
      }
    };
  }

  checkFieldValidity = (value, rules) => {
    let isValid = true;

    if (rules.required) {
      isValid = value.trim() !== "" && isValid;
    }

    if (rules.requiredDate) {
      isValid = moment.isMoment(value);
    }

    return isValid;
  };

  onInputChange = event => {
    const { handleInputChange } = this.props;
    const formFields = { ...this.state.formFields };
    const checkFieldValidity = this.checkFieldValidity;
    event.preventDefault();

    const updatedFormState = handleInputChange(
      event,
      formFields,
      checkFieldValidity
    );
    this.setState({
      ...this.state,
      ...updatedFormState
    });
  };

  onDateChange = (event, key) => {
    const { handleDateChange } = this.props;
    const formFields = { ...this.state.formFields };

    const updatedFormState = handleDateChange(
      event,
      key,
      formFields,
      this.checkFieldValidity
    );

    this.setState({
      ...this.state,
      ...updatedFormState
    });
  };

  handleSubmit = event => {
    const { generateFormData, onSubmitData } = this.props;
    event.preventDefault();

    const formData = generateFormData(this.state.formFields);

    console.log(formData);
    // onSubmitData(formData);
  };

  componentDidMount() {
    const { match, checkFormValidity } = this.props;

    const isFormValid = checkFormValidity(
      this.state.formFields,
      this.checkFieldValidity
    );

    if (match.path === "/accounts/:id/edit" && isFormValid) {
      this.setState({
        ...this.state,
        isFormValid: true
      });
    }
  }

  render() {
    const { error, isLoading, match } = this.props;
    const formTitle =
      match.path === "/transactions/new"
        ? "Add Transaction"
        : "Edit Transaction";

    const formFields = this.state.formFields;
    const inputList = (
      <Fragment>
        <SelectWrapper>
          <Label>
            {"Type"} <ColoredMark>*</ColoredMark>
          </Label>
          <Input
            elementConfig={formFields.transactionType.elementConfig}
            elementType={formFields.transactionType.elementType}
            name={"transactionType"}
            value={formFields.transactionType.value}
            onChange={this.onInputChange}
          />
        </SelectWrapper>
        <InputWrapper>
          <Label>
            {"Description"} <ColoredMark>*</ColoredMark>
          </Label>
          <Input
            elementConfig={formFields.description.elementConfig}
            elementType={formFields.description.elementType}
            name={"description"}
            value={formFields.description.value}
            touched={formFields.description.touched}
            isValid={formFields.description.isValid}
            onChange={this.onInputChange}
          />
        </InputWrapper>
        <Row>
          <InputWrapper>
            <Label>
              {"Value"} <ColoredMark>*</ColoredMark>
            </Label>
            <Input
              elementConfig={formFields.value.elementConfig}
              elementType={formFields.value.elementType}
              name={"value"}
              value={formFields.value.value}
              touched={formFields.value.touched}
              isValid={formFields.value.isValid}
              onChange={this.onInputChange}
            />
          </InputWrapper>
          <InputWrapper>
            <Label>
              {"Date"} <ColoredMark>*</ColoredMark>
            </Label>
            <StyledDatePicker
              elementConfig={formFields.date.elementConfig}
              elementType={formFields.date.elementType}
              name={"date"}
              selected={formFields.date.value}
              isValid={formFields.date.isValid}
              touched={formFields.date.touched}
              onChange={event => {
                this.onDateChange(event, "date");
              }}
            />
          </InputWrapper>
        </Row>
        <Row>
          <SelectWrapper>
            <Label>
              {"Account"} <ColoredMark>*</ColoredMark>
            </Label>
            <Input
              elementConfig={formFields.account.elementConfig}
              elementType={formFields.account.elementType}
              name={"account"}
              value={formFields.account.value}
              onChange={this.onInputChange}
            />
          </SelectWrapper>
          <SelectWrapper>
            <Label>
              {"Category"} <ColoredMark>*</ColoredMark>
            </Label>
            <Input
              elementConfig={formFields.category.elementConfig}
              elementType={formFields.category.elementType}
              name={"category"}
              value={formFields.category.value}
              onChange={this.onInputChange}
            />
          </SelectWrapper>
        </Row>
      </Fragment>
    );

    return (
      <Layout>
        <Wrapper>
          <Title>{formTitle}</Title>
          {error && <ErrorMessage>{error}</ErrorMessage>}
          <FormContainer onSubmit={this.handleSubmit}>
            {inputList}
            <SubmitRow>
              <SubmitButton
                // disabled={!this.state.isFormValid}
                onClick={this.handleSubmit}
              >
                {isLoading ? <Spinner size={25} color={"#ddd"} /> : "Submit"}
              </SubmitButton>
              <Message>
                <ColoredMark>*</ColoredMark>
                {" represent required fields."}
              </Message>
            </SubmitRow>
          </FormContainer>
        </Wrapper>
      </Layout>
    );
  }
}

Form.propTypes = {
  // TODO: Add transaction properties
  checkFormValidity: PropTypes.func.isRequired,
  error: PropTypes.string,
  generateFormData: PropTypes.func.isRequired,
  handleInputChange: PropTypes.func.isRequired,
  handleDateChange: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  match: PropTypes.object.isRequired,
  onSubmitData: PropTypes.func.isRequired
};

export default withFormHandler(Form);
