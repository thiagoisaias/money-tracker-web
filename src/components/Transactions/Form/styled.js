import styled from "styled-components";
import { devices } from "utils/devices";

import Select from "react-select";
import "assets/react-select-override.css";

import DatePicker from "react-datepicker";
import "assets/react-datepicker-override.css";

export const Wrapper = styled.div`
  background-color: #fff;
  box-shadow: 1px;
  margin: 0 auto;
  color: #484848;
  box-shadow: 0.5px 1px 1px 1px #ddd;

  @media ${devices.small} {
    padding: 32px 16px;
  }

  @media ${devices.mediumUp} {
    padding: 32px 32px 55px 32px;
    margin: 32px auto;
    border-radius: 2px;
    width: 70%;
  }
`;

export const SelectWrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 24px;
  width: 45%;
`;

export const InputWrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 24px;

  @media ${devices.small} {
    width: 100%;
  }

  @media ${devices.mediumUp} {
    width: ${props => (props.largeHalf ? "45%" : "100%")};
  }
`;

export const StyledDatePicker = styled(DatePicker)`
  border: ${props =>
    !props.isValid && props.touched ? "1px solid #ddd" : "1px solid #eee"};
  border-radius: 2px;
  font: inherit;
  color: inherit;
  cursor: pointer;
  height: 30px;
  padding-left: 12px;
  width: 100%;

  &:focus {
    border: 1px solid #ddd;
  }
`;

export const StyledSelect = styled(Select)`
  outline: none;
`;

export const StyledInput = styled.input`
  border: ${props =>
    !props.isValid && props.touched ? "1px solid #ddd" : "1px solid #eee"};
  border-radius: 2px;
  font: inherit;
  color: inherit;
  height: 30px;
  padding-left: 12px;

  &:focus {
    border: 1px solid #ddd;
  }

  &::placeholder {
    opacity: 0.6;
  }
`;

export const Row = styled.div`
  display: flex;
  position: relative;
  align-items: center;
  justify-content: space-between;
  flex-wrap: wrap;
`;

export const Message = styled.div`
  font-size: 12px;
  color: #777;
`;

export const ColoredMark = styled.span`
  font-size: 14px;
  font-weight: 600;
  color: #add8e6;
`;

export const Title = styled.h2`
  font-size: 24px;
  font-weight: 600;
  letter-spacing: 0.5px;
  margin-bottom: 32px;
`;

export const FormContainer = styled.form``;

export const Label = styled.label`
  font-weight: 400;
  color: #777;
  margin-bottom: 6px;
`;

export const SubmitRow = styled.div`
  display: flex;
  position: relative;
  align-items: center;
  justify-content: space-between;
  margin-top: 24px;
`;

export const SubmitButton = styled.button`
  border-radius: 2px;
  font: inherit;
  font-size: 12px;
  background-color: #333;
  color: #fff;
  width: 130px;
  height: 40px;

  background-color: ${props => (props.disabled ? "#777" : "#333")};

  &:hover {
    cursor: ${props => (props.disabled ? "not-allowed" : "pointer")};
    opacity: ${props => (props.disabled ? 1 : 0.95)};
  }
`;

export const ErrorMessage = styled.p`
  color: salmon;
  font-size: 14px;
  margin-top: 16px;
`;
