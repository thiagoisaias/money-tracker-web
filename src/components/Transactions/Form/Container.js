import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import { accountType, categoryType, transactionType } from "types";

import {
  clearTransactionsError,
  clearTransactionToEdit,
  createTransaction,
  updateTransaction
} from "store/actions/transactions/transactions";

import { fetchAccounts } from "store/actions/accounts/accounts";
import { fetchCategories } from "store/actions/categories/categories";

import Form from "./Form";

export class Container extends Component {
  static propTypes = {
    accountList: PropTypes.arrayOf(accountType),
    accountListLoading: PropTypes.bool.isRequired,
    categoryList: PropTypes.arrayOf(categoryType),
    categoryListLoading: PropTypes.bool.isRequired,
    transactionToEdit: transactionType,
    error: PropTypes.string,
    history: PropTypes.object.isRequired,
    isLoading: PropTypes.bool.isRequired,
    match: PropTypes.object.isRequired,
    clearTransactionError: PropTypes.func.isRequired,
    clearTransactionToEdit: PropTypes.func,
    createTransaction: PropTypes.func.isRequired,
    fetchAccounts: PropTypes.func.isRequired,
    fetchCategories: PropTypes.func.isRequired,
    updateTransaction: PropTypes.func.isRequired
  };

  onSubmitData = formData => {
    const {
      history,
      isLoading,
      match,
      createTransaction,
      updateTransaction,
      transactionToEdit
    } = this.props;

    if (isLoading) {
      return;
    }

    const { accountId, ...transformedData } = formData;

    if (match.path === "/transactions/new") {
      createTransaction(transformedData, accountId, history);
    } else if (match.path === "/transactions/:id/edit" && transactionToEdit) {
      updateTransaction(formData, transactionToEdit.id, accountId, history);
    } else {
      return;
    }
  };

  handleSelectOptions() {
    const { accountList, categoryList } = this.props;

    const accountOptionList = accountList.map(item => {
      return { value: item.id, label: item.name };
    });

    const categoryOptionList = categoryList.map(item => {
      return { value: item.id, label: item.name };
    });

    return { accountOptionList, categoryOptionList };
  }

  componentDidMount() {
    const {
      accountList,
      categoryList,
      history,
      match,
      fetchAccounts,
      fetchCategories,
      transactionToEdit
    } = this.props;

    if (categoryList.length === 0) {
      fetchCategories();
    }

    if (accountList.length === 0) {
      fetchAccounts();
    }

    if (match.path === "/transactions/:id/edit" && transactionToEdit === null) {
      history.push("/");
    }
  }

  componentWillUnmount() {
    const {
      error,
      match,
      clearTransactionError,
      clearTransactionToEdit
    } = this.props;

    if (match.path === "/transactions/:id/edit") {
      clearTransactionToEdit();
    }

    if (error) {
      clearTransactionError();
    }
  }

  render() {
    const {
      accountListLoading,
      categoryListLoading,
      error,
      isLoading,
      match,
      transactionToEdit
    } = this.props;

    const {
      accountOptionList,
      categoryOptionList
    } = this.handleSelectOptions();

    return (
      <Form
        accountOptionList={accountOptionList}
        accountListLoading={accountListLoading}
        categoryOptionList={categoryOptionList}
        categoryListLoading={categoryListLoading}
        error={error}
        isLoading={isLoading}
        match={match}
        onSubmitData={this.onSubmitData}
        transactionToEdit={transactionToEdit}
      />
    );
  }
}

Container.propTypes = {};

const mapStateToProps = state => ({
  accountList: state.accounts.accountList,
  accountListLoading: state.accounts.isLoading,
  categoryList: state.categories.categoryList,
  categoryListLoading: state.categories.isLoading,
  error: state.transactions.error,
  isLoading: state.transactions.isLoading,
  transactionToEdit: state.transactions.transactionToEdit
});

const mapDispatchToProps = {
  clearTransactionsError,
  clearTransactionToEdit,
  createTransaction,
  fetchAccounts,
  fetchCategories,
  updateTransaction
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Container)
);
