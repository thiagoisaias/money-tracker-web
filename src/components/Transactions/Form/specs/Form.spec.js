import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Form from "../Form";

describe("Transaction Form", () => {
  const mockProps = {
    accountOptionList: [
      { value: 1, label: "Account 1" },
      { value: 2, label: "Account 2" }
    ],
    accountListLoading: false,
    categoryOptionList: [
      { value: 1, label: "Food" },
      { value: 2, label: "Home" }
    ],
    categoryListLoading: false,
    checkFormValidity: jest.fn(),
    error: null,
    generateFormData: jest.fn(),
    handleInputChange: jest.fn(),
    isLoading: false,
    match: {},
    onSubmitData: jest.fn(),
    transactionData: {
      id: 1,
      account: {
        id: 1,
        name: "Lorem",
        initialBalance: 3242
      },
      date: "05-29-2018",
      description: "Lorem Ipsum",
      category: {
        id: 1,
        name: "Home",
        color: "salmon"
      },
      value: 324.32,
      transactionType: "expense"
    }
  };
  const component = shallow(<Form {...mockProps} />);

  it("renders properly", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
