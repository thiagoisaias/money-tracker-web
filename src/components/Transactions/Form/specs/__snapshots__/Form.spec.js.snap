// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction Form renders properly 1`] = `
<Form
  accountListLoading={false}
  accountOptionList={
    Array [
      Object {
        "label": "Account 1",
        "value": 1,
      },
      Object {
        "label": "Account 2",
        "value": 2,
      },
    ]
  }
  categoryListLoading={false}
  categoryOptionList={
    Array [
      Object {
        "label": "Food",
        "value": 1,
      },
      Object {
        "label": "Home",
        "value": 2,
      },
    ]
  }
  checkFormValidity={[Function]}
  error={null}
  generateFormData={[Function]}
  handleInputChange={[Function]}
  isLoading={false}
  match={Object {}}
  onSubmitData={[Function]}
  transactionToEdit={
    Object {
      "account": Object {
        "id": 1,
        "initialBalance": "3242",
        "name": "Lorem",
      },
      "category": Object {
        "color": "salmon",
        "id": 1,
        "name": "Home",
      },
      "date": "05-29-2018",
      "description": "Lorem Ipsum",
      "id": 1,
      "transactionType": "expense",
      "value": "324.32",
    }
  }
/>
`;
